File: python/solutions/cuttingthenecklace.py
import sys
from collections import deque

n, k = [int(i) for i in input().strip().split(" ")]
weights = deque([int(i) for i in input().strip().split(" ")])

total_weight = sum(weights)
if total_weight % n != 0:
    print("NO")
    sys.exit()
weight_per_person = total_weight // n

for _ in range(len(weights)):
    current_weight = 0
    possible = True
    for weight in weights:
        current_weight += weight
        if current_weight > weight_per_person:
            possible = False
            break
        if current_weight == weight_per_person:
            current_weight = 0
    if possible and current_weight == 0:
        print("YES")
        sys.exit()
    weights.rotate()
print("NO")

File: c++/solutions/cuttingthenecklace.cpp
#include <iostream>
#include <deque>

int main() {
  int n, k;
  std::deque<int> weights;
  std::cin >> n >> k;
  for (int i=0; i<k; i++) {
    int w;
    std::cin >> w;
    weights.push_front(w);
  }

  int total_weight = 0;
  for (int weight : weights) {
    total_weight += weight;
  }

  if (total_weight % n != 0) {
    std::cout << "NO" << std::endl;
    return 0;
  }

  int weight_per_person = total_weight / n;
  for (int i=0; i<k; i++) {
    int current_weight = 0;
    bool possible = true;
    for (int weight : weights) {
      current_weight += weight;
      if (current_weight > weight_per_person) {
        possible = false;
        break;
      }
      if (current_weight == weight_per_person) {
        current_weight = 0;
      }
    }
    if (possible && current_weight == 0) {
      std::cout << "YES" << std::endl;
      return 0;
    }
    weights.push_front(weights.back());
    weights.pop_back();
  }
  std::cout << "NO" << std::endl;
  return 0;
}
